doc/designs/evp-cipher-pipeline.md

```diff
@@ -0,0 +1,175 @@
+EVP APIs for supporting cipher pipelining in provided ciphers
+=============================================================
+
+OpenSSL previously supported "pipeline" ciphers via ENGINE implementations. That support was lost when we moved to providers. This document discusses API design to restore that capability and enable providers to implement such ciphers.
+
+Pipeline operation
+-------------------
+
+Certain ciphers, such as AES-GCM, can be optimized by computing blocks in parallel. Cipher pipelining support allows application to submit multiple chunks of data in one cipher update call, thereby allowing the provided implementation to take advantage of parallel computing. This is very beneficial for hardware accelerators as pipeline amortizes the latency over multiple chunks. Our libssl makes use of pipeline as discussed in [here](https://www.openssl.org/docs/manmaster/man3/SSL_CTX_set_max_pipelines.html).
+
+Pipelining with ENGINE
+-----------------------
+
+Before discussing API design for providers, let's take a look at existing pipeline API that works with engines.
+
+**EVP Interface:**
+flag to denote pipeline support
+```
+cipher->flags & EVP_CIPH_FLAG_PIPELINE
+```
+
+Input/output and aad buffers are set using `EVP_CIPHER_CTX_ctrl()`
+```
+EVP_CIPHER_CTX_ctrl() 
+    - EVP_CTRL_AEAD_TLS1_AAD (loop: one aad at a time)
+    - EVP_CTRL_SET_PIPELINE_OUTPUT_BUFS (array of buffer pointers)
+    - EVP_CTRL_SET_PIPELINE_INPUT_BUFS (array of buffer pointers)
+    - EVP_CTRL_SET_PIPELINE_INPUT_LENS
+```
+
+Single-call cipher invoked to perform encryption/decryption. 
+```
+EVP_Cipher()
+```
+
+
+Proposal for EVP pipeline APIs
+-------------------------------------
+
+Current API design is made similar to non-pipeline counterpart. The below proposal will be updated as per decisions made in next section (Design decisions).
+
+**EVP Interface:**
+API to check for pipeline support in provided cipher.
```

Comment:
With the introduction of this new API, there will be two APIs for pipelining available until the legacy code is phased out:
1. When an Engine that supports pipelining is loaded, it will set the `ctx->flags & EVP_CIPH_FLAG_PIPELINE`. If this flag is set, applications can continue to use the old API for pipelining.
2. When a Provider that supports pipelining is fetched, `EVP_CIPHER_can_pipeline()` will return true, allowing applications to utilize the new API for pipelining.  

In addition, we can map legacy API calls to the fetched provider, similar to how we currently map `EVP_CIPHER_CTX_ctrl()` to `evp_do_ciph_ctx_getparams()`. However, this may be complex since the two APIs differ significantly. It might be needed to adapt the new API to accept buffers in `OSSL_PARAM` to facilitate this translation.